File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/command/insulin/program/util/ProgramBasalUtil.java
package info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.insulin.program.util;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.insulin.program.BasalInsulinProgramElement;
import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.insulin.program.BasalShortInsulinProgramElement;
import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.insulin.program.CurrentBasalInsulinProgramElement;
import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.insulin.program.CurrentSlot;
import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.insulin.program.ShortInsulinProgramElement;
import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.definition.BasalProgram;
import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.util.MessageUtil;

public final class ProgramBasalUtil {
    public static final int MAX_DELAY_BETWEEN_TENTH_PULSES_IN_USEC_AND_USECS_IN_BASAL_SLOT = 1_800_000_000;

    public static final byte NUMBER_OF_BASAL_SLOTS = 48;
    public static final byte MAX_NUMBER_OF_SLOTS_IN_INSULIN_PROGRAM_ELEMENT = 16;

    private ProgramBasalUtil() {
    }

    public interface BasalInsulinProgramElementFactory<T extends BasalInsulinProgramElement> {
        T create(byte startSlotIndex, byte numberOfSlots, short totalTenthPulses);
    }

    public static List<BasalInsulinProgramElement> mapTenthPulsesPerSlotToLongInsulinProgramElements(short[] tenthPulsesPerSlot) {
        return mapTenthPulsesPerSlotToLongInsulinProgramElements(tenthPulsesPerSlot, BasalInsulinProgramElement::new);
    }

    public static <T extends BasalInsulinProgramElement> List<BasalInsulinProgramElement> mapTenthPulsesPerSlotToLongInsulinProgramElements(short[] tenthPulsesPerSlot, BasalInsulinProgramElementFactory<T> insulinProgramElementFactory) {
        if (tenthPulsesPerSlot.length > NUMBER_OF_BASAL_SLOTS) {
            throw new IllegalArgumentException("Basal program must contain at most 48 slots");
        }

        List<BasalInsulinProgramElement> elements = new ArrayList<>();
        long previousTenthPulsesPerSlot = 0;
        byte numberOfSlotsInCurrentElement = 0;
        byte startSlotIndex = 0;

        for (int i = 0; i < tenthPulsesPerSlot.length; i++) {
            if (i == 0) {
                previousTenthPulsesPerSlot = tenthPulsesPerSlot[i];
                numberOfSlotsInCurrentElement = 1;
            } else if (previousTenthPulsesPerSlot != tenthPulsesPerSlot[i] || (numberOfSlotsInCurrentElement + 1) * previousTenthPulsesPerSlot > 65_534) {
                elements.add(insulinProgramElementFactory.create(startSlotIndex, numberOfSlotsInCurrentElement, (short) (previousTenthPulsesPerSlot * numberOfSlotsInCurrentElement)));

                previousTenthPulsesPerSlot = tenthPulsesPerSlot[i];
                numberOfSlotsInCurrentElement = 1;
                startSlotIndex += numberOfSlotsInCurrentElement;
            } else {
                numberOfSlotsInCurrentElement++;
            }
        }
        elements.add(insulinProgramElementFactory.create(startSlotIndex, numberOfSlotsInCurrentElement, (short) (previousTenthPulsesPerSlot * numberOfSlotsInCurrentElement)));

        return elements;
    }

    public static List<ShortInsulinProgramElement> mapPulsesPerSlotToShortInsulinProgramElements(short[] pulsesPerSlot) {
        if (pulsesPerSlot.length > NUMBER_OF_BASAL_SLOTS) {
            throw new IllegalArgumentException("Basal program must contain at most 48 slots");
        }

        List<ShortInsulinProgramElement> elements = new ArrayList<>();
        boolean extraAlternatePulse = false;
        short previousPulsesPerSlot = 0;
        byte numberOfSlotsInCurrentElement = 0;
        byte currentTotalNumberOfSlots = 0;

        while (currentTotalNumberOfSlots < pulsesPerSlot.length) {
            if (currentTotalNumberOfSlots == 0) {
                // First slot

                previousPulsesPerSlot = pulsesPerSlot[0];
                currentTotalNumberOfSlots++;
                numberOfSlotsInCurrentElement = 1;
            } else if (pulsesPerSlot[currentTotalNumberOfSlots] == previousPulsesPerSlot) {
                // Subsequent slot in element (same pulses per slot as previous slot)

                if (numberOfSlotsInCurrentElement < MAX_NUMBER_OF_SLOTS_IN_INSULIN_PROGRAM_ELEMENT) {
                    numberOfSlotsInCurrentElement++;
                } else {
                    elements.add(new BasalShortInsulinProgramElement(numberOfSlotsInCurrentElement, previousPulsesPerSlot, extraAlternatePulse));
                    previousPulsesPerSlot = pulsesPerSlot[currentTotalNumberOfSlots];
                    numberOfSlotsInCurrentElement = 1;
                    extraAlternatePulse = false;
                }

                currentTotalNumberOfSlots++;
            } else if (numberOfSlotsInCurrentElement == 1 && pulsesPerSlot[currentTotalNumberOfSlots] == previousPulsesPerSlot + 1) {
                // Second slot of segment with extra alternate pulse

                boolean expectAlternatePulseForNextSegment = false;
                currentTotalNumberOfSlots++;
                numberOfSlotsInCurrentElement++;
                extraAlternatePulse = true;
                while (currentTotalNumberOfSlots < pulsesPerSlot.length) {
                    // Loop rest alternate pulse segment

                    if (pulsesPerSlot[currentTotalNumberOfSlots] == previousPulsesPerSlot + (expectAlternatePulseForNextSegment ? 1 : 0)) {
                        // Still in alternate pulse segment

                        currentTotalNumberOfSlots++;
                        expectAlternatePulseForNextSegment = !expectAlternatePulseForNextSegment;

                        if (numberOfSlotsInCurrentElement < MAX_NUMBER_OF_SLOTS_IN_INSULIN_PROGRAM_ELEMENT) {
                            numberOfSlotsInCurrentElement++;
                        } else {
                            // End of alternate pulse segment (no slots left in element)

                            elements.add(new BasalShortInsulinProgramElement(numberOfSlotsInCurrentElement, previousPulsesPerSlot, extraAlternatePulse));
                            previousPulsesPerSlot = pulsesPerSlot[currentTotalNumberOfSlots];
                            numberOfSlotsInCurrentElement = 1;
                            extraAlternatePulse = false;
                            break;
                        }
                    } else {
                        // End of alternate pulse segment (new number of pulses per slot)

                        elements.add(new BasalShortInsulinProgramElement(numberOfSlotsInCurrentElement, previousPulsesPerSlot, extraAlternatePulse));
                        previousPulsesPerSlot = pulsesPerSlot[currentTotalNumberOfSlots];
                        numberOfSlotsInCurrentElement = 1;
                        extraAlternatePulse = false;
                        currentTotalNumberOfSlots++;
                        break;
                    }
                }
            } else if (previousPulsesPerSlot != pulsesPerSlot[currentTotalNumberOfSlots]) {
                // End of segment (new number of pulses per slot)
                elements.add(new BasalShortInsulinProgramElement(numberOfSlotsInCurrentElement, previousPulsesPerSlot, extraAlternatePulse));

                previousPulsesPerSlot = pulsesPerSlot[currentTotalNumberOfSlots];
                currentTotalNumberOfSlots++;
                extraAlternatePulse = false;
                numberOfSlotsInCurrentElement = 1;
            } else {
                throw new IllegalStateException("Reached illegal point in mapBasalProgramToShortInsulinProgramElements");
            }
        }

        elements.add(new BasalShortInsulinProgramElement(numberOfSlotsInCurrentElement, previousPulsesPerSlot, extraAlternatePulse));

        return elements;
    }

    public static short[] mapBasalProgramToTenthPulsesPerSlot(BasalProgram basalProgram) {
        short[] tenthPulsesPerSlot = new short[NUMBER_OF_BASAL_SLOTS];
        for (BasalProgram.Segment segment : basalProgram.getSegments()) {
            for (int i = segment.getStartSlotIndex(); i < segment.getEndSlotIndex(); i++) {
                tenthPulsesPerSlot[i] = (short) (roundToHalf(segment.getPulsesPerHour() / 2.0d) * 10);
            }
        }

        return tenthPulsesPerSlot;
    }

    private static double roundToHalf(double d) {
        return (double) (short) ((short) (int) (d * 10.0d) / 5 * 5) / 10.0d;
    }

    public static short[] mapBasalProgramToPulsesPerSlot(BasalProgram basalProgram) {
        short[] pulsesPerSlot = new short[NUMBER_OF_BASAL_SLOTS];
        for (BasalProgram.Segment segment : basalProgram.getSegments()) {
            boolean remainingPulse = false;
            for (int i = segment.getStartSlotIndex(); i < segment.getEndSlotIndex(); i++) {
                pulsesPerSlot[i] = (short) (segment.getPulsesPerHour() / 2);
                if (segment.getPulsesPerHour() % 2 == 1) { // Do extra alternate pulse
                    if (remainingPulse) {
                        pulsesPerSlot[i] += 1;
                    }
                    remainingPulse = !remainingPulse;
                }
            }
        }

        return pulsesPerSlot;
    }

    public static CurrentSlot calculateCurrentSlot(short[] pulsesPerSlot, Date currentTime) {
        Calendar instance = Calendar.getInstance();
        instance.setTime(currentTime);

        int hourOfDay = instance.get(Calendar.HOUR_OF_DAY);
        int minuteOfHour = instance.get(Calendar.MINUTE);
        int secondOfMinute = instance.get(Calendar.SECOND);

        byte index = (byte) ((hourOfDay * 60 + minuteOfHour) / 30);
        int secondOfDay = secondOfMinute + hourOfDay * 3_600 + minuteOfHour * 60;

        short secondsRemaining = (short) ((index + 1) * 1_800 - secondOfDay);
        short pulsesRemaining = (short) ((double) pulsesPerSlot[index] * secondsRemaining / 1_800);

        return new CurrentSlot(index, (short) (secondsRemaining * 8), pulsesRemaining);
    }

    public static CurrentBasalInsulinProgramElement calculateCurrentLongInsulinProgramElement(List<BasalInsulinProgramElement> elements, Date currentTime) {
        Calendar instance = Calendar.getInstance();
        instance.setTime(currentTime);

        int hourOfDay = instance.get(Calendar.HOUR_OF_DAY);
        int minuteOfHour = instance.get(Calendar.MINUTE);
        int secondOfMinute = instance.get(Calendar.SECOND);

        int secondOfDay = secondOfMinute + hourOfDay * 3_600 + minuteOfHour * 60;
        int startSlotIndex = 0;

        byte index = 0;
        for (BasalInsulinProgramElement element : elements) {
            int startTimeInSeconds = startSlotIndex * 1_800;
            int endTimeInSeconds = startTimeInSeconds + element.getNumberOfSlots() * 1_800;

            if (secondOfDay >= startTimeInSeconds && secondOfDay < endTimeInSeconds) {
                long totalNumberOfTenThousandthPulsesInSlot = element.getTotalTenthPulses() * 1_000;
                if (totalNumberOfTenThousandthPulsesInSlot == 0) {
                    totalNumberOfTenThousandthPulsesInSlot = element.getNumberOfSlots() * 1_000;
                }

                int durationInSeconds = endTimeInSeconds - startTimeInSeconds;
                int secondsPassedInCurrentSlot = secondOfDay - startTimeInSeconds;
                long remainingTenThousandthPulses = (long) ((durationInSeconds - secondsPassedInCurrentSlot) / (double) durationInSeconds * totalNumberOfTenThousandthPulsesInSlot);
                int delayBetweenTenthPulsesInUsec = (int) (durationInSeconds * 1_000_000L * 1_000 / totalNumberOfTenThousandthPulsesInSlot);
                int secondsRemaining = secondsPassedInCurrentSlot % 1_800;
                int delayUntilNextTenthPulseInUsec = delayBetweenTenthPulsesInUsec;
                for (int i = 0; i < secondsRemaining; i++) {
                    delayUntilNextTenthPulseInUsec = delayUntilNextTenthPulseInUsec - 1_000_000;
                    while (delayUntilNextTenthPulseInUsec <= 0) {
                        delayUntilNextTenthPulseInUsec += delayBetweenTenthPulsesInUsec;
                    }
                }
                short remainingTenthPulses = (short) ((remainingTenThousandthPulses % 1_000 != 0 ? 1 : 0) + remainingTenThousandthPulses / 1_000);

                return new CurrentBasalInsulinProgramElement(index, delayUntilNextTenthPulseInUsec, remainingTenthPulses);
            }

            index++;
            startSlotIndex += element.getNumberOfSlots();
        }
        throw new IllegalStateException("Could not determine current long insulin program element");
    }

    public static short calculateChecksum(short[] pulsesPerSlot, CurrentSlot currentSlot) {
        ByteBuffer buffer = ByteBuffer.allocate(1 + 2 + 2 + NUMBER_OF_BASAL_SLOTS * 2) //
                .put(currentSlot.getIndex()) //
                .putShort(currentSlot.getPulsesRemaining()) //
                .putShort(currentSlot.getEighthSecondsRemaining());

        for (short pulses : pulsesPerSlot) {
            buffer.putShort(pulses);
        }

        return MessageUtil.calculateChecksum(buffer.array());
    }
}

File: pump/omnipod-dash/src/main/kotlin/app/aaps/pump/omnipod/dash/driver/pod/command/insulin/program/util/ProgramBasalUtil.kt
package app.aaps.pump.omnipod.dash.driver.pod.command.insulin.program.util

import app.aaps.pump.omnipod.dash.driver.pod.command.insulin.program.BasalInsulinProgramElement
import app.aaps.pump.omnipod.dash.driver.pod.command.insulin.program.BasalShortInsulinProgramElement
import app.aaps.pump.omnipod.dash.driver.pod.command.insulin.program.CurrentBasalInsulinProgramElement
import app.aaps.pump.omnipod.dash.driver.pod.command.insulin.program.CurrentSlot
import app.aaps.pump.omnipod.dash.driver.pod.command.insulin.program.ShortInsulinProgramElement
import app.aaps.pump.omnipod.dash.driver.pod.definition.BasalProgram
import app.aaps.pump.omnipod.dash.driver.pod.util.MessageUtil
import java.nio.ByteBuffer
import java.util.*

object ProgramBasalUtil {

    const val MAX_DELAY_BETWEEN_TENTH_PULSES_IN_USEC_AND_USECS_IN_BASAL_SLOT = 1800000000
    private const val NUMBER_OF_BASAL_SLOTS: Byte = 48
    private const val MAX_NUMBER_OF_SLOTS_IN_INSULIN_PROGRAM_ELEMENT: Byte = 16

    fun mapTenthPulsesPerSlotToLongInsulinProgramElements(
        tenthPulsesPerSlot: ShortArray,
        insulinProgramElementFactory: (Byte, Byte, Short) -> BasalInsulinProgramElement = ::BasalInsulinProgramElement
    ): List<BasalInsulinProgramElement> {
        require(tenthPulsesPerSlot.size <= NUMBER_OF_BASAL_SLOTS) { "Basal program must contain at most 48 slots" }
        val elements: MutableList<BasalInsulinProgramElement> = ArrayList()
        var previousTenthPulsesPerSlot: Short = 0
        var numberOfSlotsInCurrentElement: Byte = 0
        var startSlotIndex: Byte = 0
        for (i in tenthPulsesPerSlot.indices) {
            if (i == 0) {
                previousTenthPulsesPerSlot = tenthPulsesPerSlot[i]
                numberOfSlotsInCurrentElement = 1
            } else if (previousTenthPulsesPerSlot != tenthPulsesPerSlot[i] || (numberOfSlotsInCurrentElement + 1) * previousTenthPulsesPerSlot > 65534) {
                elements.add(
                    insulinProgramElementFactory(
                        startSlotIndex,
                        numberOfSlotsInCurrentElement,
                        (previousTenthPulsesPerSlot * numberOfSlotsInCurrentElement).toShort()
                    )
                )
                previousTenthPulsesPerSlot = tenthPulsesPerSlot[i]
                numberOfSlotsInCurrentElement = 1
                startSlotIndex = (numberOfSlotsInCurrentElement + startSlotIndex).toByte()
            } else {
                numberOfSlotsInCurrentElement++
            }
        }
        elements.add(
            insulinProgramElementFactory(
                startSlotIndex,
                numberOfSlotsInCurrentElement,
                (previousTenthPulsesPerSlot * numberOfSlotsInCurrentElement).toShort()
            )
        )
        return elements
    }

    fun mapPulsesPerSlotToShortInsulinProgramElements(pulsesPerSlot: ShortArray?): List<ShortInsulinProgramElement> {
        require(pulsesPerSlot!!.size <= NUMBER_OF_BASAL_SLOTS) { "Basal program must contain at most 48 slots" }

        val elements: MutableList<ShortInsulinProgramElement> = ArrayList()
        var extraAlternatePulse = false
        var previousPulsesPerSlot: Short = 0
        var numberOfSlotsInCurrentElement: Byte = 0
        var currentTotalNumberOfSlots: Byte = 0
        while (currentTotalNumberOfSlots < pulsesPerSlot.size) {
            if (currentTotalNumberOfSlots.toInt() == 0) {
                // First slot
                previousPulsesPerSlot = pulsesPerSlot[0]
                currentTotalNumberOfSlots++
                numberOfSlotsInCurrentElement = 1
            } else if (pulsesPerSlot[currentTotalNumberOfSlots.toInt()] == previousPulsesPerSlot) {
                // Subsequent slot in element (same pulses per slot as previous slot)
                if (numberOfSlotsInCurrentElement < MAX_NUMBER_OF_SLOTS_IN_INSULIN_PROGRAM_ELEMENT) {
                    numberOfSlotsInCurrentElement++
                } else {
                    elements.add(
                        BasalShortInsulinProgramElement(
                            numberOfSlotsInCurrentElement,
                            previousPulsesPerSlot,
                            extraAlternatePulse
                        )
                    )
                    previousPulsesPerSlot = pulsesPerSlot[currentTotalNumberOfSlots.toInt()]
                    numberOfSlotsInCurrentElement = 1
                    extraAlternatePulse = false
                }
                currentTotalNumberOfSlots++
            } else if (numberOfSlotsInCurrentElement.toInt() == 1 && !extraAlternatePulse && pulsesPerSlot[currentTotalNumberOfSlots.toInt()].toInt() == previousPulsesPerSlot + 1) {
                // Second slot of segment with extra alternate pulse
                var expectAlternatePulseForNextSegment = false
                currentTotalNumberOfSlots++
                numberOfSlotsInCurrentElement++
                extraAlternatePulse = true
                while (currentTotalNumberOfSlots < pulsesPerSlot.size) {
                    // Loop rest alternate pulse segment
                    if (pulsesPerSlot[currentTotalNumberOfSlots.toInt()].toInt() == previousPulsesPerSlot + (if (expectAlternatePulseForNextSegment) 1 else 0)) {
                        // Still in alternate pulse segment
                        expectAlternatePulseForNextSegment = !expectAlternatePulseForNextSegment
                        if (numberOfSlotsInCurrentElement < MAX_NUMBER_OF_SLOTS_IN_INSULIN_PROGRAM_ELEMENT) {
                            numberOfSlotsInCurrentElement++
                            currentTotalNumberOfSlots++
                        } else {
                            // End of alternate pulse segment (no slots left in element)
                            elements.add(
                                BasalShortInsulinProgramElement(
                                    numberOfSlotsInCurrentElement,
                                    previousPulsesPerSlot,
                                    extraAlternatePulse
                                )
                            )
                            previousPulsesPerSlot = pulsesPerSlot[currentTotalNumberOfSlots.toInt()]
                            numberOfSlotsInCurrentElement = 1
                            extraAlternatePulse = false
                            currentTotalNumberOfSlots++
                            break
                        }
                    } else {
                        // End of alternate pulse segment (new number of pulses per slot)
                        elements.add(
                            BasalShortInsulinProgramElement(
                                numberOfSlotsInCurrentElement,
                                previousPulsesPerSlot,
                                extraAlternatePulse
                            )
                        )
                        previousPulsesPerSlot = pulsesPerSlot[currentTotalNumberOfSlots.toInt()]
                        numberOfSlotsInCurrentElement = 1
                        extraAlternatePulse = false
                        currentTotalNumberOfSlots++
                        break
                    }
                }
            } else if (previousPulsesPerSlot != pulsesPerSlot[currentTotalNumberOfSlots.toInt()]) {
                // End of segment (new number of pulses per slot)
                elements.add(
                    BasalShortInsulinProgramElement(
                        numberOfSlotsInCurrentElement,
                        previousPulsesPerSlot,
                        extraAlternatePulse
                    )
                )
                previousPulsesPerSlot = pulsesPerSlot[currentTotalNumberOfSlots.toInt()]
                currentTotalNumberOfSlots++
                extraAlternatePulse = false
                numberOfSlotsInCurrentElement = 1
            } else {
                throw IllegalStateException("Reached illegal point in mapBasalProgramToShortInsulinProgramElements")
            }
        }
        elements.add(
            BasalShortInsulinProgramElement(
                numberOfSlotsInCurrentElement,
                previousPulsesPerSlot,
                extraAlternatePulse
            )
        )
        return elements
    }

    fun mapBasalProgramToTenthPulsesPerSlot(basalProgram: BasalProgram): ShortArray {
        val tenthPulsesPerSlot = ShortArray(NUMBER_OF_BASAL_SLOTS.toInt())
        for (segment in basalProgram.segments) {
            for (i in segment.startSlotIndex until segment.endSlotIndex) {
                tenthPulsesPerSlot[i] = (roundToHalf(segment.getPulsesPerHour() / 2.0) * 10).toInt()
                    .toShort() // TODO Adrian: int conversion ok?
            }
        }
        return tenthPulsesPerSlot
    }

    private fun roundToHalf(d: Double): Double {
        return ((d * 10.0).toInt().toShort() / 5 * 5).toShort().toDouble() / 10.0
    }

    fun mapBasalProgramToPulsesPerSlot(basalProgram: BasalProgram): ShortArray {
        val pulsesPerSlot = ShortArray(NUMBER_OF_BASAL_SLOTS.toInt())
        for (segment in basalProgram.segments) {
            var remainingPulse = false
            for (i in segment.startSlotIndex until segment.endSlotIndex) {
                pulsesPerSlot[i] = (segment.getPulsesPerHour() / 2).toShort()
                if (segment.getPulsesPerHour() % 2 == 1) { // Do extra alternate pulse
                    if (remainingPulse) {
                        pulsesPerSlot[i] = (pulsesPerSlot[i] + 1).toShort()
                    }
                    remainingPulse = !remainingPulse
                }
            }
        }
        return pulsesPerSlot
    }

    fun calculateCurrentSlot(pulsesPerSlot: ShortArray?, currentTime: Date): CurrentSlot {
        val instance = Calendar.getInstance()
        instance.time = currentTime
        val hourOfDay = instance[Calendar.HOUR_OF_DAY]
        val minuteOfHour = instance[Calendar.MINUTE]
        val secondOfMinute = instance[Calendar.SECOND]

        val index = ((hourOfDay * 60 + minuteOfHour) / 30).toByte()
        val secondOfDay = secondOfMinute + hourOfDay * 3600 + minuteOfHour * 60
        val secondsRemaining = ((index + 1) * 1800 - secondOfDay).toShort()
        val pulsesRemaining = (pulsesPerSlot!![index.toInt()].toDouble() * secondsRemaining / 1800).toInt().toShort()
        return CurrentSlot(index, (secondsRemaining * 8).toShort(), pulsesRemaining)
    }

    fun calculateCurrentLongInsulinProgramElement(
        elements: List<BasalInsulinProgramElement>,
        currentTime: Date
    ): CurrentBasalInsulinProgramElement {
        val instance = Calendar.getInstance()
        instance.time = currentTime
        val hourOfDay = instance[Calendar.HOUR_OF_DAY]
        val minuteOfHour = instance[Calendar.MINUTE]
        val secondOfMinute = instance[Calendar.SECOND]
        val secondOfDay = secondOfMinute + hourOfDay * 3600 + minuteOfHour * 60
        var startSlotIndex = 0
        var index: Byte = 0
        for (element in elements) {
            val startTimeInSeconds = startSlotIndex * 1800
            val endTimeInSeconds = startTimeInSeconds + element.numberOfSlots * 1800
            if (secondOfDay in startTimeInSeconds until endTimeInSeconds) { // TODO Adrian Range check ok
                var totalNumberOfTenThousandthPulsesInSlot = (element.totalTenthPulses * 1000).toLong()
                if (totalNumberOfTenThousandthPulsesInSlot == 0L) {
                    totalNumberOfTenThousandthPulsesInSlot = (element.numberOfSlots * 1000).toLong()
                }
                val durationInSeconds = endTimeInSeconds - startTimeInSeconds
                val secondsPassedInCurrentSlot = secondOfDay - startTimeInSeconds
                val remainingTenThousandthPulses =
                    ((durationInSeconds - secondsPassedInCurrentSlot) / durationInSeconds.toDouble() * totalNumberOfTenThousandthPulsesInSlot).toLong()
                val delayBetweenTenthPulsesInUsec =
                    (durationInSeconds * 1000000L * 1000 / totalNumberOfTenThousandthPulsesInSlot).toInt()
                val secondsRemaining = secondsPassedInCurrentSlot % 1800
                var delayUntilNextTenthPulseInUsec = delayBetweenTenthPulsesInUsec
                for (i in 0 until secondsRemaining) {
                    delayUntilNextTenthPulseInUsec -= 1000000
                    while (delayUntilNextTenthPulseInUsec <= 0) {
                        delayUntilNextTenthPulseInUsec += delayBetweenTenthPulsesInUsec
                    }
                }
                val remainingTenthPulses =
                    ((if (remainingTenThousandthPulses % 1000 != 0L) 1 else 0) + remainingTenThousandthPulses / 1000).toShort()
                return CurrentBasalInsulinProgramElement(index, delayUntilNextTenthPulseInUsec, remainingTenthPulses)
            }
            index++
            startSlotIndex += element.numberOfSlots.toInt()
        }
        throw IllegalStateException("Could not determine current long insulin program element")
    }

    fun calculateChecksum(pulsesPerSlot: ShortArray?, currentSlot: CurrentSlot): Short {
        val buffer = ByteBuffer.allocate(1 + 2 + 2 + NUMBER_OF_BASAL_SLOTS * 2)
            .put(currentSlot.index)
            .putShort(currentSlot.pulsesRemaining)
            .putShort(currentSlot.eighthSecondsRemaining)
        for (pulses in pulsesPerSlot!!) {
            buffer.putShort(pulses)
        }
        return MessageUtil.calculateChecksum(buffer.array())
    }
}
